crates/uv/tests/it/version.rs

```diff
@@ -0,0 +1,1137 @@
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    version = "1.10.31"
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Dynamic version should error on read
+#[test]
+fn version_get_dynamic() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myproject"
+        dynamic = ["version"]
+        requires-python = ">=3.12"
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version(), @r"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: There is no 'project.version' field in: pyproject.toml
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    dynamic = ["version"]
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Dynamic version should error on write
+#[test]
+fn version_set_dynamic() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myproject"
+        dynamic = ["version"]
+        requires-python = ">=3.12"
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version()
+        .arg("0.1.2"), @r"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: There is no 'project.version' field in: pyproject.toml
+    ");
+
+    let pyproject = fs_err::read_to_string(&pyproject_toml)?;
+    assert_snapshot!(
+        pyproject,
+    @r#"
+    [project]
+    name = "myproject"
+    dynamic = ["version"]
+    requires-python = ">=3.12"
+    "#
+    );
+    Ok(())
+}
+
+// Should fallback to `uv --version` if this pyproject.toml isn't usable for whatever reason
+// (In this case, because tool.uv.managed = false)
+#[test]
+fn version_get_fallback() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "myapp"
+        version = "0.1.2"
+        
+        [tool.uv]
+        managed = false
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.version(), @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    uv [VERSION] ([COMMIT] DATE)
+
+    ----- stderr -----
+    warning: failed to read project: The project is marked as unmanaged: `[TEMP_DIR]/`
+      running `uv self version` for compatibility with old `uv version` command.
+      this fallback will be removed soon, pass `--project .` to make this an error.
```

Comment:
Reading this... I wonder if we should suggest `--preview` to make this an error? `--project .` is a weird thing for someone to keep around later?

Reply:
Conversely `--preview` will enable a ton of random other things right..? Maybe that's fine though.